test/topology/test_topology.py

```diff
@@ -112,6 +114,45 @@ async def test_replace_reuse_ip(manager: ManagerClient, random_tables) -> None:
     # TODO: check that group 0 no longer contains the replaced node (#12153)
 
 
+# Checks basic functionality on the cluster with different values of the --smp parameter on the nodes.
+@pytest.mark.asyncio
+async def test_nodes_with_different_smp(manager: ManagerClient, random_tables: RandomTables) -> None:
```

Comment:
So the test relies on a bug in the testing framework which allows starting two independent Scylla clusters under a single test framework cluster. (There should be 1-1 correspondence between framework's `ScyllaCluster` and an actual Scylla cluster). It basically works by accident.

Please modify the test so it doesn't depend on this bug.

Wouldn't it reproduce if:
- we didn't stop the existing servers,
- we added 2 new servers to the existing servers with --smp 3 and --smp 4; so in total we'd end up with 5 servers: 3 of them with --smp 2, one with --smp 3, and one with --smp 4. We have three different values of `smp` in this cluster, should be enough to reproduce the bug
- perform `add_tables` with this 5-node cluster. Maybe do it a few times to increase reproducibility.

Optionally we could decommission one of the original servers so that we have 4 servers in the cluster, forcing every quorum to have at least two different `smp`s. The decommission operation itself could reproduce the problem too.

Reply:
Even adding the first server with a different `smp` may reproduce the problem, right?

Reply:
I've spent some time and figured out the scenario where the problem reproduces in a most stable and natural way. The scenario - to deal with a fresh cluster with different `--smp` values. I do not want to adapt my testing needs to the nuances of the current testing infrastructure. This is a wrong path. It's a testing infra which should accommodate my needs. Otherwise, the tests will be dictated by the current infra and not the real world scenarios. 

The problem is that tests get the predefined cluster with existing nodes. Maybe it's what most tests want (I'm not sure), but it's not what is needed here. So I want to replace this cluster entirely. I can mark this action with some specific api in ScyllaCluster (`replace_cluster` or something) if that helps.

>Wouldn't it reproduce the targeted problem if:
> we didn't stop the existing servers,
> we added 2 new servers to the existing servers with --smp 3 and --smp 4; so in total we'd end up with 5 servers: 3 of them with --smp 2, one with --smp 3, and one with --smp 4. We have three different values of smp in this cluster, should be enough to reproduce the bug
> perform add_tables with this 5-node cluster. Maybe do it a few times to increase reproducibility.

I tried this scenario and various other, they were less stable to reproduce the problem.

Reply:
> I do not want to adapt my testing needs to the nuances of the current testing infrastructure. This is a wrong path. It's a testing infra which should accommodate my needs. Otherwise, the tests will be dictated by the current infra and not the real world scenarios.

I agree that the framework is not flexible enough and needs to be extended. But it won't extend itself. Someone has to do it. Almost all tests in test_topology.py required extending the framework when they were written.

Right now, for a suite, there is a `Pool` of clusters used by all tests in the suite. The parameters of the cluster server are defined statically by the suite in a yaml file. This is quite limiting as you observed. For example, to test Raft upgrade scenarios I created a separate suite which starts clusters with Raft initially disabled.

It would be much better if each test case could specify the parameters of the cluster that it needs, and the pool would populate itself with clusters according to the needs of the collected test cases. More flexible, although it has downsides too, like less reusability of clusters between tests (although in topology tests currently we never reuse clusters because all become "dirty" after topology changes) and potentially much more complicated implementation of pool/test scheduler etc.

As a first step maybe we could allow tests to ask for empty clusters (using a special marker or fixture or fixture parameter?), so a test like this would have to start the cluster manually. Not sure how hard that would be to implement. May require a separate `Pool` of empty clusters or maybe we can mix the clusters in the existing `Pool` (probably bad idea, would lead to other tests obtaining these empty clusters as well).

> I tried this scenario and various other, they were less stable to reproduce the problem.

How frequently does the current approach reproduce the problem compared to e.g. what I proposed?